File: BackendApi/Application/Services/SyncService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Application.DTOs;
using Application.Interfaces;
using Entities.Models;
using Infra.Context;
using Microsoft.EntityFrameworkCore;

namespace Application.Services;

public class SyncService : ISyncService
{
    private readonly ColetaDbContext _context;
    public SyncService(ColetaDbContext context) => _context = context;

    public async Task SincronizarLoteAsync(IEnumerable<ColetaSyncDto> coletas, Guid idUsuario)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();
        try
        {
            foreach (var dto in coletas)
            {
                // Idempotencia: ignora registros ja sincronizados
                if (await _context.RegistrosColeta.AnyAsync(r => r.IdRegistro == dto.IdRegistro))
                    continue;

                // 1. Familia
                var familia = await _context.Familias.FirstOrDefaultAsync(f => f.IdFamilia == dto.IdFamilia);
                if (familia == null)
                {
                    familia = new Familia
                    {
                        IdFamilia             = dto.IdFamilia,
                        CodigoFamilia         = $"FAM-SYNC-{dto.IdFamilia.ToString()[..8]}",
                        Endereco              = dto.Endereco,
                        Bairro                = dto.Bairro,
                        Comunidade            = dto.Comunidade,
                        QtdMoradores          = dto.QtdMoradores,
                        RendaFamiliarMensal   = dto.RendaFamiliarMensal,
                        RecebeBeneficioSocial = dto.RecebeBeneficioSocial,
                        BeneficioSocial       = dto.BeneficioSocial,
                        PossuiInternetCasa    = dto.PossuiInternetCasa,
                        TipoAcessoInternet    = dto.TipoAcessoInternet,
                        CriadoPor             = idUsuario,
                        CreatedAt             = DateTime.UtcNow,
                        UpdatedAt             = DateTime.UtcNow
                    };
                    _context.Familias.Add(familia);
                }

                // 2. Aluno
                var aluno = await _context.Alunos.FirstOrDefaultAsync(a => a.IdAluno == dto.IdAluno);
                if (aluno == null)
                {
                    aluno = new Aluno
                    {
                        IdAluno    = dto.IdAluno,
                        IdFamilia  = familia.IdFamilia,
                        NomeAluno  = dto.NomeAluno,
                        DataNascimento = dto.DataNascimento,
                        Sexo       = dto.Sexo,
                        CpfAluno   = dto.CpfAluno,
                        NecessidadeEducacionalEspecial = dto.NecessidadeEducacionalEspecial,
                        DescricaoNecessidade = dto.DescricaoNecessidade,
                        CreatedAt  = DateTime.UtcNow,
                        UpdatedAt  = DateTime.UtcNow
                    };
                    _context.Alunos.Add(aluno);
                }

                // 3. Responsavel
                var responsavel = await _context.Responsaveis.FirstOrDefaultAsync(r => r.CpfResponsavel == dto.CpfResponsavel);
                if (responsavel == null)
                {
                    responsavel = new Responsavel
                    {
                        IdResponsavel      = Guid.NewGuid(),
                        NomeResponsavel    = dto.NomeResponsavel,
                        CpfResponsavel     = dto.CpfResponsavel,
                        TelefoneResponsavel = dto.TelefoneResponsavel,
                        EmailResponsavel   = dto.EmailResponsavel,
                        CreatedAt          = DateTime.UtcNow
                    };
                    _context.Responsaveis.Add(responsavel);
                }

                // 4. Vinculo Aluno-Responsavel
                bool relExiste = await _context.AlunosResponsaveis
                    .AnyAsync(ar => ar.IdAluno == aluno.IdAluno && ar.IdResponsavel == responsavel.IdResponsavel);
                if (!relExiste)
                {
                    _context.AlunosResponsaveis.Add(new AlunoResponsavel
                    {
                        IdAluno              = aluno.IdAluno,
                        IdResponsavel        = responsavel.IdResponsavel,
                        ParentescoResponsavel = dto.ParentescoResponsavel
                    });
                }

                // 5. Matricula
                _context.Matriculas.Add(new Matricula
                {
                    IdMatricula          = Guid.NewGuid(),
                    IdAluno              = aluno.IdAluno,
                    AnoLetivo            = DateTime.UtcNow.Year,
                    AnoSerie             = dto.AnoSerie,
                    Turno                = dto.Turno,
                    FrequenciaEscolarPct = dto.FrequenciaEscolarPct,
                    MeioTransporteEscola = dto.MeioTransporteEscola,
                    TempoDeslocamentoMin = dto.TempoDeslocamentoMin
                });

                // 6. Registro de coleta
                _context.RegistrosColeta.Add(new RegistroColeta
                {
                    IdRegistro          = dto.IdRegistro,
                    IdAluno             = aluno.IdAluno,
                    IdFamilia           = familia.IdFamilia,
                    IdUsuario           = idUsuario,
                    Observacao          = dto.Observacao,
                    DataColeta          = DateTime.UtcNow,
                    StatusSincronizacao = "SINCRONIZADO",
                    SincronizadoEm      = DateTime.UtcNow
                });

                await _context.SaveChangesAsync();
            }

            await transaction.CommitAsync();
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }
    }
}

File: mobile-app/lib/data/sync_service.dart
import 'package:sqflite/sqflite.dart';
import '../core/database_helper.dart';
import 'api_client.dart';
import 'models/coleta_sync_dto.dart';

class SyncService {
  final ApiClient _apiClient = ApiClient();
  final DatabaseHelper _dbHelper = DatabaseHelper.instance;

  Future<void> syncPendingRecords() async {
    final db = await _dbHelper.database;
    
    // Pega os registros pendentes
    final registros = await db.query(
      'registros_coleta',
      where: 'statusSincronizacao = ?',
      whereArgs: ['PENDENTE'],
    );

    if (registros.isEmpty) return;

    List<Map<String, dynamic>> payload = [];

    for (var registro in registros) {
      final idRegistro = registro['idRegistro'] as String;
      final idAluno = registro['idAluno'] as String;
      final idFamilia = registro['idFamilia'] as String;
      final observacao = registro['observacao'] as String?;

      // Busca Familia
      final familiaMap = (await db.query('familias', where: 'idFamilia = ?', whereArgs: [idFamilia])).first;
      
      // Busca Aluno
      final alunoMap = (await db.query('alunos', where: 'idAluno = ?', whereArgs: [idAluno])).first;

      // Busca Vínculo Aluno-Responsável
      final vinculoMap = (await db.query('alunos_responsaveis', where: 'idAluno = ?', whereArgs: [idAluno])).first;
      final idResponsavel = vinculoMap['idResponsavel'] as String;
      
      // Busca Responsável
      final responsavelMap = (await db.query('responsaveis', where: 'idResponsavel = ?', whereArgs: [idResponsavel])).first;

      // Busca Matrícula (pode não existir em registros antigos)
      final matriculaList = await db.query('matriculas', where: 'idAluno = ?', whereArgs: [idAluno]);
      final matriculaMap = matriculaList.isNotEmpty ? matriculaList.first : null;

      final dto = ColetaSyncDto(
        idRegistro: idRegistro,
        idFamilia: idFamilia,
        idAluno: idAluno,
        nomeAluno: alunoMap['nomeAluno'] as String,
        dataNascimento: alunoMap['dataNascimento'] as String,
        sexo: alunoMap['sexo'] as String? ?? '',
        cpfAluno: alunoMap['cpfAluno'] as String? ?? '',
        necessidadeEducacionalEspecial: (alunoMap['necessidadeEducacionalEspecial'] as int? ?? 0) == 1,
        descricaoNecessidade: alunoMap['descricaoNecessidade'] as String? ?? '',
        nomeResponsavel: responsavelMap['nomeResponsavel'] as String,
        parentescoResponsavel: vinculoMap['parentescoResponsavel'] as String,
        cpfResponsavel: responsavelMap['cpfResponsavel'] as String? ?? '',
        telefoneResponsavel: responsavelMap['telefoneResponsavel'] as String? ?? '',
        emailResponsavel: responsavelMap['emailResponsavel'] as String? ?? '',
        endereco: familiaMap['endereco'] as String,
        bairro: familiaMap['bairro'] as String,
        comunidade: familiaMap['comunidade'] as String? ?? '',
        qtdMoradores: familiaMap['qtdMoradores'] as int? ?? 0,
        rendaFamiliarMensal: (familiaMap['rendaFamiliarMensal'] as num?)?.toDouble() ?? 0.0,
        recebeBeneficioSocial: (familiaMap['recebeBeneficioSocial'] as int? ?? 0) == 1,
        beneficioSocial: familiaMap['beneficioSocial'] as String? ?? '',
        possuiInternetCasa: (familiaMap['possuiInternetCasa'] as int? ?? 0) == 1,
        tipoAcessoInternet: familiaMap['tipoAcessoInternet'] as String? ?? '',
        meioTransporteEscola: matriculaMap?['meioTransporteEscola'] as String? ?? '',
        tempoDeslocamentoMin: matriculaMap?['tempoDeslocamentoMin'] as int? ?? 0,
        frequenciaEscolarPct: (matriculaMap?['frequenciaEscolarPct'] as num?)?.toDouble() ?? 0.0,
        anoSerie: matriculaMap?['anoSerie'] as String? ?? '',
        turno: matriculaMap?['turno'] as String? ?? '',
        observacao: observacao ?? '',
      );

      payload.add(dto.toJson());
    }

    try {
      await _apiClient.postList('/sync', payload);
      
      // Se sucesso, atualiza o status para SINCRONIZADO no banco local
      Batch batch = db.batch();
      for (var p in payload) {
        batch.update(
          'registros_coleta',
          {'statusSincronizacao': 'SINCRONIZADO'},
          where: 'idRegistro = ?',
          whereArgs: [p['idRegistro']],
        );
      }
      await batch.commit(noResult: true);
    } catch (e) {
      print('Erro ao sincronizar: $e');
      rethrow;
    }
  }
}
